components/psst/browser/core/psst_rule_registry.cc

```diff
@@ -29,75 +30,88 @@ namespace psst {
-  return {test_script, policy_script, version};
+std::optional<MatchedRule> CreateMatchedRule(
+    base::WeakPtr<PsstRuleRegistryImpl> registry,
+    RuleDataReader* rule_data_reader,
+    const PsstRule& rule) {
+  if (!registry) {
+    return std::nullopt;
+  }
+  return MatchedRuleFactory(rule_data_reader, rule.Name(), rule.Version())
+      .Create(rule);
 }
 
 }  // namespace
 
 // static
-PsstRuleRegistry* PsstRuleRegistry::GetInstance() {
-  // Check if feature flag is enabled.
+PsstRuleRegistryAccessor* PsstRuleRegistryAccessor::GetInstance() {
+  return base::Singleton<PsstRuleRegistryAccessor>::get();
+}
+
+PsstRuleRegistryAccessor::PsstRuleRegistryAccessor()
+    : impl_(std::make_unique<PsstRuleRegistryImpl>()) {}
+PsstRuleRegistryAccessor::~PsstRuleRegistryAccessor() = default;
+
+PsstRuleRegistry* PsstRuleRegistryAccessor::Registry() {
   if (!base::FeatureList::IsEnabled(psst::features::kBravePsst)) {
     return nullptr;
   }
-  return base::Singleton<PsstRuleRegistry>::get();
+
+  return impl_.get();
 }
 
-PsstRuleRegistry::PsstRuleRegistry() = default;
+void PsstRuleRegistryAccessor::SetRegistryForTesting(
+    std::unique_ptr<PsstRuleRegistry> new_inst) {
+  impl_ = std::move(new_inst);
+}
 
-PsstRuleRegistry::~PsstRuleRegistry() = default;
+PsstRuleRegistryImpl::PsstRuleRegistryImpl() = default;
+PsstRuleRegistryImpl::~PsstRuleRegistryImpl() = default;
 
-void PsstRuleRegistry::CheckIfMatch(
+void PsstRuleRegistryImpl::CheckIfMatch(
     const GURL& url,
-    base::OnceCallback<void(MatchedRule)> cb) const {
+    base::OnceCallback<void(const std::optional<MatchedRule>&)> cb) {
   for (const PsstRule& rule : rules_) {
     if (rule.ShouldInsertScript(url)) {
       base::ThreadPool::PostTaskAndReplyWithResult(
           FROM_HERE, {base::MayBlock()},
-          base::BindOnce(&CreateMatchedRule, component_path_,
-                         rule.GetTestScript(), rule.GetPolicyScript(),
-                         rule.GetVersion()),
+          base::BindOnce(&CreateMatchedRule, weak_factory_.GetWeakPtr(),
+                         rule_data_reader_.get(), rule),
           std::move(cb));
       // Only ever find one matching rule.
       return;
     }
   }
 }
 
-void PsstRuleRegistry::LoadRules(const base::FilePath& path) {
-  SetComponentPath(path);
+void PsstRuleRegistryImpl::SetRuleDataReaderForTest(
+    std::unique_ptr<RuleDataReader> rule_data_reader) {
+  rule_data_reader_ = std::move(rule_data_reader);
+}
+
+void PsstRuleRegistryImpl::LoadRules(const base::FilePath& path) {
+  if (!rule_data_reader_) {
+    rule_data_reader_ = std::make_unique<RuleDataReader>(path);
+  }
+
   base::ThreadPool::PostTaskAndReplyWithResult(
       FROM_HERE, {base::MayBlock()},
       base::BindOnce(&ReadFile, path.Append(kJsonFile)),
-      base::BindOnce(&PsstRuleRegistry::OnLoadRules,
+      base::BindOnce(&PsstRuleRegistryImpl::OnLoadRules,
                      weak_factory_.GetWeakPtr()));
```

Comment:
this is also unsafe for the same reason as https://github.com/brave/brave-core/pull/29044/files#r2089899849

Reply:
oops, nvm, I read this wrong. This one is safe